TSE::matchSpectrum() 1st prototype

pcolaianni opened the pull request:
@dmccloskey For the code to work, it needs the .msp file to be put in the data folder with the name `TargetedSpectraExtractor_mainLib.MSP`.

What would you like to change to the current interface? Do the resulting scores look legit? (with reference to what I wrote on Slack)

pcolaianni committed:
feat: implement MSP format parser for TargetedSpectraExtractor::matchSpectrum()

pcolaianni committed:
feat: matchSpectrum() prototype

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

dmccloskey commented on src/openms/include/OpenMS/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.h:
```diff
@@ -191,6 +200,20 @@ namespace OpenMS
     );
 
+    /**
+      @brief Searches for the best matches for the input spectrum, against a library
```
"Searches the spectral library for the top scoring candidates that match the input spectrum."

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:
```diff
@@ -33,8 +33,8 @@
-
-#include <boost/unordered_map.hpp> 
+#include <OpenMS/COMPARISON/SPECTRA/BinnedSpectralContrastAngle.h>
+#include <boost/unordered_map.hpp>
```
Is the boost unordered_map implementation needed or would the std library C++11 implementation suffice?  If the latter, we should probably just use the std library implementation over boost now that OpenMS has moved to C++11.

pcolaianni replied:
Not needed anymore. Removing it :smiley:

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:
```diff
@@ -426,4 +435,51 @@ namespace OpenMS
+    std::map<String,double> scores_map;
+    for (const MSSpectrum& s : experiment.getSpectra())
+    {
+      if (similarity_function_ == BINNED_SPECTRAL_CONTRAST_ANGLE)
```
I will be curious to see how `binned_spectral_constrast_angle` compares to the 2nd option: "Peak matching between spectra".  Could you provide the reference for the OpenMS implementation for that like you did for the  `binned_spectral_constrast_angle`.  That was very helpful.

pcolaianni replied:
I am confused about this point. I think I have not made mention of a second option, yet. But yeah I'm writing the code in a way that we'll propose more than one choice for the similarity function. The spectral contrast angle (https://www.sciencedirect.com/science/article/pii/S1044030501003270) seems the best option as of today (I'll verify this), but I'll also try other ways.

dmccloskey replied:
You have not made mention of a second option, but it was one option mentioned by Timo when we spoke with him.  It is listed in the issue as one of the recommended methods. 

Please keep me updated on what you find in the literature.  If you need help reviewing the different methods, please let me know and I can help evaluate them as well.

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:
```diff
@@ -426,4 +435,51 @@ namespace OpenMS
+        const BinnedSpectrum bs2 (s, bin_size, false, 2); // NOTE: magic value
+        BinnedSpectralContrastAngle cmp;
+        scores_map.insert( {s.getName(), cmp(bs1, bs2)} );
+      }
```
Let's discuss the "magic values" for the bin sizes more during our meeting.  These maybe parameters best left for the user to define in the `Params` object when running the method.  However, it would be beneficial for us to suggest some good starting values for the user.

dmccloskey commented on src/openms/source/ANALYSIS/OPENSWATH/TargetedSpectraExtractor.cpp:
```diff
@@ -426,4 +435,51 @@ namespace OpenMS
+    );
+
+    // Output best N matches
+    const Size N { 5 };
```
We should make the number of top N hits a parameter called `top_scores_to_report` for the user to specify.

dmccloskey commented on src/openms/source/FORMAT/MSPTSEFile.cpp:
```diff
@@ -0,0 +1,152 @@
+// --------------------------------------------------------------------------
+
+#include <OpenMS/FORMAT/MSPTSEFile.h>
+#include <boost/regex.hpp>
```
Lets change this to the std library implementation now that C++11 is supported by OpenMS.  See https://theboostcpplibraries.com/boost.regex

pcolaianni replied:
The current version of the compilers installed on the build machine is too old to provide all the new regex functionalities. The code compiles (with some small changes to the regular expressions, which should not be necessary once they update the compilers), but the feature implementations have problems. Once AIT support replies, I'll verify this point.

dmccloskey replied:
OK thanks for looking into this with them.  We definitely want a compiler that supports all C++11 features at a minimum.

pcolaianni replied:
I confirm that things worked out without the boost implementation. The code is now present in commit: https://github.com/biosustain/OpenMS/pull/150/commits/f6a90f3ec77a448ca8817a0f555837c68eb23e2a

dmccloskey commented on src/openms/source/FORMAT/MSPTSEFile.cpp:
```diff
@@ -0,0 +1,152 @@
+        MSSpectrum::StringDataArrays& SDAs = spectrum.getStringDataArrays();
+        SDAs.resize(7);
+        // TODO: part of this info could be saved as integer instead of string
+        SDAs[0].setName("Synon");
```
I am not sure if we can generalize these names to all .msp files for metabolomics.  I would imagine for instance that the "NIST#" is specific only to the NIST library.  I would also imagine that they could vary quite dramatically between different spectral library databases.  I will double check on the header names for some other prominent databases in the field (e.g., Metlin, MSCloud, etc,.) to see which fields are required and which are optional.

dmccloskey commented on src/openms/source/FORMAT/MSPTSEFile.cpp:
```diff
@@ -0,0 +1,152 @@
+        spectrum.clear(true);
+        spectrum.setName( std::string(m[1]) );
+        LOG_DEBUG << spectrum.getName() << std::endl;
+        MSSpectrum::StringDataArrays& SDAs = spectrum.getStringDataArrays();
```
We may need to infer the array size by the first entry in the .msp file if they number of entries is different per spectral library database.

dmccloskey commented on src/tests/class_tests/openms/source/MSPTSEFile_test.cpp:
```diff
@@ -0,0 +1,120 @@
+MSPTSEFile* ptr = nullptr;
+MSPTSEFile* null_ptr = nullptr;
+// const String input_filepath = OPENMS_GET_TEST_DATA_PATH("TargetedSpectraExtractor_mainLib.MSP");
+const String input_filepath = OPENMS_GET_TEST_DATA_PATH("MSPTSEFile_input.msp");
```
Be sure that we do not include the whole file...just a small subset for the tests.

dmccloskey commented on src/tests/class_tests/openms/source/TargetedSpectraExtractor_test.cpp:
```diff
@@ -104,6 +105,8 @@ TargetedSpectraExtractor* ptr = 0;
 TargetedSpectraExtractor* null_ptr = 0;
 const String experiment_path = OPENMS_GET_TEST_DATA_PATH("TargetedSpectraExtractor_13C1_spectra0to100.mzML");
 const String target_list_path = OPENMS_GET_TEST_DATA_PATH("TargetedSpectraExtractor_13CFlux_TraML.csv");
+const String msp_path = OPENMS_GET_TEST_DATA_PATH("TargetedSpectraExtractor_mainLib.MSP");
```
Lets just use a subset of the data for when we are ready to push to OpenMS

dmccloskey reviewed (approved):
This looks quite promising for a first implementation.  Great work.

pcolaianni committed:
docs: update docs (review)

feat: add top_matches_to_report class parameter

chore: remove <boost/unordered_map.hpp> include

chore: use <regex> instead of boost version

test: isolate matchSpectrum tests from others (ie. contants, params, etc)

Tests currently fail because of some metadata missing from experiment file

pcolaianni commented:
@dmccloskey _TSEMSPFile_ is still assuming that metadata exist in a certain shape. I'll update that part once we know how MSP files usually look. I will probably change the code in a way that only the necessary information is assumed to be present. The rest is going to be dynamically available through `getStringDataArrayByName()` or similar methods.

Do you have a better name for the parser class? _TSEMSPFile_ doesn't look good imho. (I also noticed I actually called it _MSPTSEFile_ ... renaming it)

pcolaianni committed:
chore: Rename MSPTSEFile to TseMSPFile

pcolaianni committed:
chore: 18 extracted spectra

pcolaianni committed:
fix: refine params (gauss width and snr)

pcolaianni committed:
feat: avoid adding duplicates in msp library

feat: add "Num Peaks" metadata

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
feat: implement `pushParsedInfoToNamedDataArray()`

pcolaianni committed:
docs: Add small comments to new method's if/else

dmccloskey commented on src/openms/include/OpenMS/FORMAT/TseMSPFile.h:
```diff
@@ -49,17 +49,17 @@ namespace OpenMS
     > 136 350; 
   */
-  class OPENMS_DLLAPI MSPTSEFile
+  class OPENMS_DLLAPI TseMSPFile
```
I think something like "MSPMetaboFile" might be more appropriate.

dmccloskey commented on src/openms/source/FORMAT/TseMSPFile.cpp:
```diff
@@ -0,0 +1,183 @@
+    while (!ifs.eof())
+    {
+      ifs.getline(line, BUFSIZE);
+      // TODO: Should metadata all be saved as strings?
```
I think that saving the metadata as a string is the best option for now.

dmccloskey commented on src/openms/source/FORMAT/TseMSPFile.cpp:
```diff
@@ -0,0 +1,183 @@
+    std::regex re_synon("^Synon: (.+)");
+    std::regex re_formula("^Formula: (.+)");
+    std::regex re_mw("^MW: (.+)");
+    std::regex re_cas_nist("^CAS#: (.+);  NIST#: (.+)");
```
We could search using `regex` for the "mandatory" fields as implemented, but it might be more robust for the variable fields to have some way of pulling them out based on the document structure (i.e., we know that all fields will come before the individual spectrum).  That way we will not miss any fields that we could not foresee ahead of time.

dmccloskey replied:
We should also make a note in the code which fields are "mandatory" and which ones are "optional".

pcolaianni replied:
Good idea. I'll try to implement that.

pcolaianni replied:
The problem I see with this implementation is that, for example, in our exported file we have two metadata in one line (`CAS#` and `NIST#`).
I will assume that a metadatum ends at a new line or at a `;` character.

dmccloskey reviewed (commented):
The code is moving in the right direction.  I made a few suggestions that I think could help the generalization of the MSP parser as well as a few other suggestions.

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
feat: implement addSpectrumToExperiment()

feat: implement getStringDataArrayByName()

feat: make loaded_spectra_names_ a class member

test: add friend class to test or use private/protected methods

test: print synonyms

pcolaianni committed:
style: rename TseMSPFile to MSPMetaboFile

pcolaianni committed:
feat: parse generic metadata

pcolaianni commented:
@dmccloskey Everything from comments is implemented.
What should happen if a parsed spectrum misses the `Comments` and/or `Num Peaks` metadata? Throw an exception?

dmccloskey commented:
@pasqoo lets just throw an exception in that case.  We can re-evaluate that choice when we get more use cases.

pcolaianni committed:
fix: set `adding_spectrum` to false when a duplicate is found

feat: implement check on Name

pcolaianni commented:
@dmccloskey Implemented. But I had to disable the check on `Comments` because it wouldn't pass the tests (NIST library has some spectra without said field).
`Num Peaks` is present in all spectra within the library, though.

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

dmccloskey commented:
@pasqoo let's set comments as optional as well then but we should set a default empty string if no comment is provided so that the output is gauranteed to have a comment field.

pcolaianni committed:
Merge branch 'develop' into feature/TSE_MatchSpectrum

pcolaianni committed:
feat: ensure Comments metadatum is present

pcolaianni committed:
test: verify that absent Comments is an empty string

pcolaianni commented:
@dmccloskey done

pcolaianni committed:
feat: implement class params bin_size and peak_spread

test: display results differently
